### ISSUE TITLE
Pyright can't infer the type of a repeated list

### ISSUE THREAD

**Issue by refi64** [microsoft/pyright#736]
**Describe the bug**

Given code like this:

```python
x: List[Optional[int]] = [None] * 2
```

Pyright will assume the list on the right hand side is `List[None]` despite the type annotation. [Mypy can, however](https://mypy-play.net/?mypy=latest&python=3.8&gist=d6918459cdf07afd6ce5d99c0021579e).

**To Reproduce**

- Copy paste the example into VS Code.
- See the red squiggly
- Profit (?)

**Expected behavior**

The code should type check.

**Screenshots or Code**

Already posted above.

**VS Code extension or command-line**

VS Code extension 1.1.43.

**Additional context**

I'm guessing it's that pyright doesn't see the RHS as a constant expression or similar because of the `*`?

**Comment by erictraut** [comment]
This behavior is "as designed".

Pyright is able to do "bidirectional inference", using the expected type of the LHS to guide the type inference of the RHS. It supports this for assignments where the RHS is a list expression, dictionary expression, set expression, and a comprehension. It does not handle arbitrary expressions though.

In your example, the RHS is a binary expression (using the `*` operator). The behavior of this operator is defined by the `__mul__` method in the `list` class. Pyright doesn't have any intrinsic knowledge about the operations supported by the `list` class. It needs to consult the builtins.pyi type stub to understand what methods are supported. To evaluate the return type of the `__mul__` method, it needs to first know the specialized type of the list (i.e. it needs to determine the type of the type argument `_T` in the generic type `List[_T]`). So it needs to infer the type of the expression `[None]` before it can determine the type of the expression `[None] * 2`. As you would expect, the inferred type of `[None]` is `List[None]`, so the return type of the `__mul__` method is the same type.

I suspect that mypy doesn't complain in this case because it's a bit loose in some of its type consistency checks. In other words, you're probably benefitting from a bug (or a type-checking blind spot) in mypy.

If my supposition is incorrect, then it means mypy has hard-coded some knowledge about the `list` class rather than relying on the stdlib type stubs for that information. That's possible, but it would be a pretty ugly hack, and I've tried to avoid doing that in Pyright.

Here are some techniques that will work with Pyright without generating a type error:
```
x: List[Optional[int]] = [None, None]
```
```
xx: List[Optional[int]] = [None]
x = xx * 2
```
```
x: Sequence[Optional[int]] = [None] * 2
```

You can also use a `# type: ignore` comment to suppress the error if you really want to stick with your existing formulation.

**Comment by refi64** [comment]
I'm not aware of mypy having anything for that, though given what you said I did play around a bit more, and I think it's just that mypy always tries to apply the LHS's type into more complex expressions that just collection literals, e.g. this works in mypy because it's able to infer `T` as having to be `Optional[int]` but fails in pyright:

```python
T = TypeVar('T')

def func(x: List[T]) -> List[T]:
    return x

x: List[Optional[int]] = func([None])
```

That being said, I'd imagine this can be slower, so I understand that this is probably just a design decision.

**Comment by erictraut** [comment]
This isn't just a question of analysis speed or a design decision. There's a chicken-and-egg problem when bidirectional inference is used to infer the return type of a method call on an object that's instantiated from a generic class. (Bidirectional inference means that we're using the "expected type" to inform the inferred type of an expression.) The reasoning behind this is not obvious, so at the risk of boring you with too many details, let me explain.

Here are the steps that Pyright follows when it encounters the code in your original example at the top of this thread.

1. Determine the "expected type" by evaluating the type of the LHS of the assignment. In your example, this type is provided by your annotation: `List[Optional[int]]`.
2. Infer the type of the expression on the RHS of the assignment using the "expected type" as a guide. In your original example above, the expression on the RHS is `[None] * 2`, which is effectively `[None].__mul__(2)`.
3. Before we can determine the return type of the `__mul__` method, we need to infer the type of `[None]`. This is an instance of the `List` class, but `List` is generic. We need to specialize it before we can evaluate `__mul__(2)`. The type checker can't assume that this object type has anything to do with the expected return type of the method, so it needs to perform this inference without the benefit of the "expected type" we evaluated above. In this case, we need to infer that the type is `List[None]`.
4. Now that we have a specialized object type, we can look up the `__mul__` method. In this case, it's defined as `def __mul__(self, n: int) -> List[_T]: ...` in builtins.pyi.
5. Next, we need to "bind" that method to the object. This involves partial specialization (replacing type variables) within the method signature and dropping the "self" parameter. The result is a function of type `def __mul__(n: int) -> List[None]`. Note that the TypeVar `_T` has been replaced by `None`.
6. Finally, we can use the "expected type" computed in step 1 to infer the return type of the method. In this case, there's no inference necessary because we have a declared and specialized type already. So the result is type `List[None]`.

The chicken-and-egg problem I mentioned above is because of step 3, which is required when dealing with a method.

In the case of a simple function (as opposed to a method), we don't need to perform steps 3 or 5, so this chicken-and-egg problem doesn't exist.

In your second example `func` is a simple function, so this should work fine. I've updated Pyright to handle your second example, but I still don't see any way to handle methods without following the steps I've listed above, so your initial example still doesn't work.

**Comment by erictraut** [comment]
Your second example now works in Pyright version 1.1.44, which I just published.

**Comment by refi64** [comment]
Thanks for the explanation! I enjoy learning stuff like this.
